crates/storage/nippy-jar/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+                                (maybe_zstd_compressors.as_mut(), compression)
+                            {
+                                zstd.compress_with_dictionary(
+                                    &value,
+                                    &mut tmp_buf,
+                                    &mut file,
+                                    Some(dict_compressors.get_mut(column_number).expect("exists")),
+                                )?;
+                            } else {
+                                compression.compress_to(&value, &mut file)?;
+                            }
+                        } else {
+                            file.write_all(&value)?;
+                        }
+                    }
+                    None => {
+                        return Err(NippyJarError::UnexpectedMissingValue(
+                            row_number,
+                            column_number as u64,
+                        ))
+                    }
+                }
+
+                iterators.push(column_iter);
+            }
+
+            row_number += 1;
+            if row_number == total_rows {
+                break
+            }
+
+            column_iterators = iterators.into_iter();
+        }
+
+        // Write offset index to file
+        bincode::serialize_into(File::create(self.index_path())?, &values_offsets)?;
+
+        Ok(())
+    }
+
+    /// Safety checks before creating and returning a [`File`] handle to write data to.
+    fn freeze_check(
+        &mut self,
+        columns: &Vec<impl IntoIterator<Item = Vec<u8>>>,
+    ) -> Result<File, NippyJarError> {
+        if columns.len() != self.columns {
+            return Err(NippyJarError::ColumnLenMismatch(self.columns, columns.len()))
+        }
+
+        if let Some(compression) = &self.compressor {
+            if !compression.is_ready() {
+                return Err(NippyJarError::CompressorNotReady)
+            }
+        }
+
+        Ok(File::create(self.data_path())?)
+    }
+
+    /// Writes all necessary configuration to file.
+    fn freeze_config(&mut self, handle: &mut File) -> Result<(), NippyJarError> {
+        // TODO Split Dictionaries and Bloomfilters Configuration so we dont have to load everything
+        // at once
+        Ok(bincode::serialize_into(handle, &self)?)
+    }
+}
+
+impl Filter for NippyJar {
+    fn add(&mut self, element: &[u8]) -> Result<(), NippyJarError> {
+        self.filter.as_mut().ok_or(NippyJarError::FilterMissing)?.add(element)
+    }
+
+    fn contains(&self, element: &[u8]) -> Result<bool, NippyJarError> {
+        self.filter.as_ref().ok_or(NippyJarError::FilterMissing)?.contains(element)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum NippyJarError {
+    #[error("err")]
+    Err,
+    #[error("err")]
+    Disconnect(#[from] std::io::Error),
+    #[error("err")]
+    Bincode(#[from] Box<bincode::ErrorKind>),
```

Comment:
these messages ignore the underlying error when displayed

Reply:
handled in upstream pr